### PR TITLE
Enhance XML Coverage report files to contain tokenized source code

### PR DESCRIPTION
**Pull request by theseer**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -29,7 +29,8 @@
         "phpunit/php-text-template": "^1.2",
         "sebastian/code-unit-reverse-lookup": "^1.0",
         "sebastian/environment": "^2.0",
-        "sebastian/version": "^2.0"
+        "sebastian/version": "^2.0",
+        "theseer/tokenizer": "^1.1.0"
```
Comment (GrahamCampbell) [inline composer.json]: Just `^1.1` is fine here?

File: src/Report/Xml/Facade.php [hunk 2]
```diff
@@ -148,7 +148,7 @@ private function processFile(FileNode $file, Directory $context)
         }
 
         foreach ($file->getCoverageData() as $line => $tests) {
-            if (!is_array($tests) || count($tests) == 0) {
+            if (!is_array($tests) || count($tests) === 0) {
```
Comment (GrahamCampbell) [inline src/Report/Xml/Facade.php]: Technically unrelated change.

File: src/Report/Xml/Source.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
+<?php declare(strict_types = 1);
+/*
+ * This file is part of the php-code-coverage package.
+ *
+ * (c) Sebastian Bergmann <sebastian@phpunit.de>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+namespace SebastianBergmann\CodeCoverage\Report\Xml;
```
Comment (GrahamCampbell) [inline src/Report/Xml/Source.php]: missing newline

File: src/Report/Xml/Source.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+use TheSeer\Tokenizer\Tokenizer;
+use TheSeer\Tokenizer\XMLSerializer;
+
+class Source
+{
+    /** @var \DOMElement */
+    private $context;
+
+    /**
+     * @param \DOMElement $context
```
Comment (GrahamCampbell) [inline src/Report/Xml/Source.php]: this phpdoc technically adds no value

File: src/Report/Xml/Source.php (hunk 3 above)
Comment (theseer) [inline src/Report/Xml/Source.php]: Not according to php-cs-fixer

File: src/Report/Xml/Source.php (hunk 4 above)
Comment (theseer) [inline src/Report/Xml/Source.php]: No shit Sherlock...

File: composer.json (hunk 1 above)
Comment (theseer) [inline composer.json]: Yes?

File: src/Report/Xml/Facade.php (hunk 2 above)
Comment (theseer) [inline src/Report/Xml/Facade.php]: *sigh*

File: src/Report/Xml/Source.php (hunk 4 above)
Comment (GrahamCampbell) [inline src/Report/Xml/Source.php]: Well, my point, was why add it?

File: src/Report/Xml/Source.php (hunk 3 above)
Comment (GrahamCampbell) [inline src/Report/Xml/Source.php]: Well, php-cs-fixer isn't total. I was just referring to the fact that in the other files in this repo, there is always an extra newline there:

image

File: composer.json (hunk 1 above)
Comment (GrahamCampbell) [inline composer.json]: Lol. True. What I really meant was, this should probably changed to that, to match the rest of the file.

Comment (theseer) [inline composer.json]: Ah. Okay, point taken.

File: src/Report/Xml/Source.php (hunk 4 above)
Comment (theseer) [inline src/Report/Xml/Source.php]: With PHP 7.0+, PHPDoc pretty much never adds any technical value. Yet, it is still considered a good practice to have them.
